DSEGOG-62 showing/hiding plot channels

sam-glendenning opened the pull request:
## Description
This allows the user to toggle the visibility of a plotted data channel in the plot by toggling a button in its channel label. Note this does not remove the plotted data from the plot - it only toggles its opacity and removes its label from the legend, thus keeping the plot scales the same.

This attempts to replicate the behaviour in eCat. If a channel's visibility is toggled off and we export to SVG, it will export exactly what is shown in the plot, i.e. minus the channel's data. If we export the plot data, all the data in the plot will be exported, including hidden channels. This is the same as in eCat but can be changed if we want to.

## Testing instructions
Use the visibility toggle to show/hide channels in the plot. Try exporting data to test the expected results.

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage

## Agile board tracking
Closes DSEGOG-62

Commit:
DSEGOG-62 adding SelectedPlotChannel type and basic visibility button

Defining a type for housing details about selected plot channels, currently their name and visibility in plot but later their colour, etc.

Also adding a visibility on/off button in the channel label, currently does nothing

Commit:
DSEGOG-62 allowing toggle visibility of channels in plot

This does not remove the data from the plot, all it does it toggle its opacity. This ensures the plot stays at the same scales, etc.

Commit:
DSEGOG-62 aligning visibility button next to close button

Not managed to get the text wrapping yet

sam-glendenning commented on src/plotting/plotSettings.component.tsx:
```diff
@@ -512,11 +552,33 @@ const PlotSettings = (props: PlotSettingsProps) => {
-                    aria-label={`Remove ${plotChannel} axis`}
-                    onClick={() => removePlotChannel(plotChannel)}
-                  />
+                  <Typography noWrap>{plotChannel.name}</Typography>
```
This has been a pain. The text does not truncate. Therefore, if you click on a really long channel name, the entirety of the label is rendered and the width of the container element is extended to match, creating a horizontal scroll bar. I've tried to fix this by adding `xs` this and `width: 'inherit'` that but had no luck so any help would be appreciated here for style points.

louise-davies replied:
Honestly, I think the easiest solution here is to set a sensible `maxWidth` on the `Typography` based on the drawer width - everything else. Experimentally, `190px` works well

sam-glendenning replied:
Guess I was trying to be too clever! `maxWidth="190"` works very well, thank you :)

Commit:
DSEGOG-62 switching from icon to icon button

This adds a ripple effect when the button is clicked

Commit:
DSEGOG-62 adding and fixing unit tests

Commit:
DSEGOG-62 fixing broken test

Commit:
DSEGOG-62 bumping plotSettings coverage

Commit:
DSEGOG-62 bumping coverage in plot.component

Commit:
DSEGOG-62 fixing minor bug

Commit:
Fix to plotWindow unit test

This was a test that has been failing locally for me but passing on CI. The close settings button was being incorrectly detected when it shouldn't have been. Fixed this by ensuring its visibility is hidden at the right time

louise-davies commented on src/plotting/plot.component.tsx:
```diff
@@ -112,47 +122,64 @@ const Plot = (props: PlotProps) => {
-          </VictoryGroup>
-        ))}
+        {selectedChannels.map((channel) => {
+          const currentDataset = datasets.find(
```
I wonder, if we're structuring it that we have the selected channel config in one prop and the data in another - is it not easier/more efficient to have `datasets` just be a single data array, with objects containing fields equalling the channel names? Then, when creating the lines, we go through the list of `selectedChannels`, using the name as the key for the Y axis for each line and the same data array passed as `data`?

i.e. rather than passing `datasets={[{name: "shotnum", data: [{"time": 1, "shotnum":1}, {"time": 2, "shotnum":2}, {"time": 3, "shotnum":3}]}, {name: "activeArea", data: [{"time": 1, "activeArea":1}, {"time": 2, "activeArea":2}, {"time": 3, "activeArea":3}]}]}`, we pass in `data={[{"time": 1, "shotnum":1, "activeArea": 1}, {"time": 2, "shotnum":2, "activeArea": 2}, {"time": 3, "shotnum":3, "activeArea": 3}]}` (we reduce repeated X axis info this way)

Again, this isn't necessarily the best way to do things/worth refactoring but it's just a thought

sam-glendenning replied:
Possibly, I just found it easier to determine what dataset matched with what selectedChannel info by just adding a name field to the dataset, instead of having to extract this from the data itself. What do you think?

louise-davies replied:
We can keep it like this for now - we can check later on if things need changing once we do more larger scale tests

louise-davies replied:
Although in the method I propose there is no difficulty in determining the dataset - as there's only one dataset that is supplied to all the lines, and just the y axis value is adjusted

Commit:
DSEGOG-62 adding maxWidth to selectedChannel labels

This ensures text truncation. We'll need to adjust this as more icons are added to the label
